cloudimg/ms_azure.py

```diff
@@ -345,6 +346,9 @@ def upload_to_container(self, image_path, container_name, object_name,
         If the file_url starts with http, use the upload_blob_from_url method,
         otherwise use the upload_blob method.
 
+        If the incoming image is compressed in ``.xz`` (LZMA) format and it
+        doesn't come from HTTP it will upload the uncompressed content.
```

Comment:
I was wondering if this behavior difference would be a problem.  I looked into upload_blob_from_url and now I have two questions about this.  If you could explain a bit under which cases the URL/HTTP case is expected to be used it would be helpful.

1. It seems to me that upload_blob_from_url does not actually *upload* from the client at all, instead it asks the remote service to *download* from the specified URL.

Is that correct and was that behavior already understood?  I'm a bit surprised because I thought the intent with supporting both local paths, and URLs, was that you could pass e.g. a URL to an archive attached to an internal koji build.  But this wouldn't work since that service is not Internet-facing.

https://learn.microsoft.com/en-us/rest/api/storageservices/put-blob-from-url also suggests a max size of 256MiB for the blob which seems not enough for the files we've been dealing with.

If these limitations are known, it's surprising the doc string doesn't mention it at all.

2. Does the lack of decompression in the URL case make sense?

Again, I thought the idea with accepting paths or URLs here was that you could point the library to a file on local storage or a file on any accessible remote service and it'd do the right thing.  But the fact that decompression happens for one and not the other kinda breaks this.

I'm assuming that, in the .xz case, the blob *has* to be uploaded decompressed - because if not then I don't get why we're decompressing it.  In that case, calling this with a URL pointing at an .xz file would do the wrong thing - successfully uploading, but creating a useless blob.  Thus the API is easy to use incorrectly as it will seem to do the right thing for most inputs, but silently do something wrong for this kind of input.

Can you please check my logic?  If you agree that, with the current implementation of this method, calling it with the URL of an .xz file would always be wrong - then I'd suggest actually checking for that condition and raising an exception.  Otherwise the API invites bugs.

Alternatively, if the outcome of (1) is that upload_blob_from_url is not suitable and it has to be changed to really download and upload the file, then it seems like supporting decompression also in the URL case would be straightforward.

Reply:
You're right, the `upload_blob_from_url` requires us to expose a public URL so their remote can download the file from. This is not what I intended when writing the library so in this case I believe this call could be completely removed.

With this I'm inclined to 2 different approches:

1. Remove the `upload_from_blob_url` from `cloudimg` as it won't be used since we don't expose a public URL and it has a size limitation that I was not aware of before.
2. Keep the URL and raise an error if the file comes as `.xz`

The last alternative you pointed out of downloading from URL and uploading it is not what `upload_blob_from_url` does and even if it was it would require some tmp storage, which is something we're trying to avoid.

With this, can you please let me know which approaches do you prefer? Personally I prefer the 1 to completely drop the `upload_from_url` but I'm not sure if this goes against the retrocompatiblity or not... Finally I'll adjust the code according to the best approach.

Thanks once again for the good catch!

Reply:
I don't know for sure why the URL support is here in the first place.

It might be because of a discussion we had earlier in the project.  At that time, Pub was not yet in openshift and the method of reading content from brew after the openshift migration was still an open question.  I remember suggesting that supporting the load of content both via a plain open() or by fetching a remote file would be wise since it was possible we were going to migrate all content loading to HTTP.

In the time since then, that matter was settled and you can assume the usual NFS mounts are available.  So, if that was the only reason to support URLs in the first place, you could just drop it.

Personally I feel that supporting URLs is still of some value because the NFS approach is not perfect (e.g. right now it uses a slightly outdated mirror) and I can see us moving away from it in the long term.  But it's your call, and it can always be reimplemented later if needed.

If URL support was kept, I believe decompressing could be supported without needing a lot of temporary storage because the HTTP client for downloading, the LZMA decompressor and the azure client for uploading all support streaming.  It should not be necessary to download the entire file and decompress it.  The only gotcha seems to be that whatever content is accessed *must* be served up with a Content-Length header, otherwise we're back at "you need to decompress all the data before you can know the length for the upload request".

Reply:
> I don't know for sure why the URL support is here in the first place.

Maybe because URL is supported in aws as well
* https://github.com/release-engineering/cloudimg/blob/18b6e67be87815739341876a8643cee18595e17b/cloudimg/aws.py#L397-L413

> otherwise we're back at "you need to decompress all the data before you can know the length for the upload request"

My plan is to ask image builder team to provide metadata in koji about size of decompressed file.
Which would allow to replace `data.seek(0, os.SEEK_END)` and thus speed up the process.
And that would allow for using URL which is not accessible publicly for azure. Just wrapping into stream
```
resp = requests.get('http://some.internal.utl/test.vhd.xz', stream=True)
raw_stream = iterable_to_stream(resp.iter_content())  # https://gist.github.com/mechanical-snail/7688353
decompressed_stream = lzma.LZMAFile(raw_stream)
```
It it will still require `upload_blob` and not `upload_blob_from_url`
But such discussion would be a bit offtopic.

For me the the current documentation is enough because it is expected by 
* https://learn.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.blobclient?view=azure-python#azure-storage-blob-blobclient-upload-blob-from-url

Checking suffix for URL and raise exception is not ideal. Because server can return `Content-type: application/x-xz` even for different name.

@rohanpm If current approach with documenting decompression just for file is not enough than I woudl agree with @JAVGan `prefer the 1 to completely drop the upload_from_url`
Unless you are against due to backward incompatible change.

Reply:
I've included a new commit to drop this feature, PTAL again and let me know if it's ok now